src/metaswitch/clearwater/config_manager/move_config.py

```diff
@@ -298,143 +463,118 @@ def validate_config(force=False):
-    # `upload_shared_config`, we check that the port we expect etcd to be
-    # listening on is actually open before doing anything else. It's possible
-    # that we don't actually need to do that in this case, because we'll
-    # always be doing some sort of initial verification that the connection
-    # can be made. But need to confirm that is in fact the case.
-
-    # TODO: if force is True, check that we are root, otherwise error out
-    # Validate the config
-    validate_config(force)
-
-    upload_config(autoconfirm, config_loader, config_type, force, local_store)
+def upload_config(autoconfirm, config_loader, config_type, force, local_store):
+    """Read the relevant config from file and upload it to etcd."""
+    remote_revision = ready_for_upload_checks(autoconfirm,
+                                              config_loader,
+                                              config_type,
+                                              local_store)
+
+    # Upload the configuration to the etcd cluster. This will trigger the
+    # queue manager to schedule nodes to be restarted.
+    config_loader.write_config_to_etcd(config_type, remote_revision)
 
+    # Clearwater can be run with multiple etcd clusters. The apply_config_key
+    # variable stores the information about which etcd cluster the changes
+    # should be applied to.
+    apply_config_key = subprocess.check_output(
+        "/usr/share/clearwater/clearwater-queue-manager/scripts/get_apply_config_key")
 
-def upload_config(autoconfirm, config_loader, config_type, force, local_store):
-    local_config, local_revision = local_store.load_config_and_revision(
-        config_type)
-    remote_config, remote_revision = config_loader.get_config_and_index(
-        config_type)
+    # If the config changes are being forced through, the queue manager needs
+    # to be made aware so it knows to push on in the case of an error when
+    # applying the config to a node.
+    subprocess.call(["/usr/share/clearwater/clearwater-queue-manager/scripts/modify_nodes_in_queue",
+                     "force_true" if force else "force_false",
+                     apply_config_key])
+
+    # If we reach this point then config upload was successful. Cleaning up
+    # the config file we've uploaded makes sure we don't cause confusion later.
+    local_store.config_cleanup()
+
+    print "{} successfully uploaded".format(shared_config)
+
+
+def ready_for_upload_checks(autoconfirm,
+                            config_loader,
+                            config_type,
+                            local_store):
+    """Make sure that we can and should upload config. Returns the current
+    revision of the master config upload should proceed, otherwise throws
+    a ConfigUploadFailed exception."""
+    try:
+        local_config, local_revision = local_store.load_config_and_revision(
+            config_type)
+    except IOError:
+        log.error("%s cannot be uploaded", config_type)
```

Comment:
This isn't the correct error - we don't know whether it could be uploaded only that we couldn't read the downloaded config from disk.